Add new OAuth flow for addons to PDvis

MichaelViveros opened the pull request:
@alisdair Please review. I tested both cases where the user generates a new OAuth token and where the user has a bad token and tries to make an API request (they get redirected to generate a new OAuth token).

Some questions:

1. What's the best way to generate the `state` [param](https://github.com/PagerDuty/addons/commit/306c8dec966eaadfb18ee23e697fc9f0d80bb7e9#diff-073f48c76203d07dc39b9549aa405ff4R30) for OAuth?

2. What's the best / most intuitive way to break up the OAuth logic in [`main`](https://github.com/PagerDuty/addons/commit/306c8dec966eaadfb18ee23e697fc9f0d80bb7e9#diff-d4165290cddbf3212be36d99822dd88aR829) so that addon devs who want to upgrade their addon to use OAuth can do so as easily as possible? I broke it up into `requestOAuthToken` and `receiveOAuthToken` methods to make the OAuth workflow easy to understand (making a request to the OAuth service, getting redirected to PagerDuty and authorizing the addon, getting redirected back to the app with the OAuth token and state).

alisdair commented on PDvis/common.js:
```diff
@@ -17,25 +20,54 @@ function getParametersByName(name) {
+  window.localStorage.setItem('pdvisClientState', client_state);
+  var client_id = "b5236b4b54e5ec7b2b51cccc603174a2ac0b575f33753e04a2924dced669c03b";
+  var redirect_uri = "http://localhost:8000";
+  var oauth_route = "http://app.pagerduty.net/oauth/authorize?client_id=" + client_id + "&redirect_uri=" + redirect_uri + "&response_type=token&state=" + client_state;
```
Should use production URLs, I think?

alisdair commented on PDvis/common.js:
```diff
@@ -17,25 +20,54 @@ function getParametersByName(name) {
+function getOAuthResponseParams() {
+  var oauth_params = [];
+	var hash = window.location.hash.replace(/^#/, '?');
+	var param_token = getParameterByName('access_token', hash);
```
Indentation inconsistency here.

alisdair commented on PDvis/vis.js:
```diff
@@ -826,6 +826,18 @@ function setProgressBar(progress) {
 }
 
 function main() {
+	debugger
```
Stray debugger

alisdair commented on PDvis/common.js:
```diff
@@ -17,25 +20,54 @@ function getParametersByName(name) {
+
+function receiveOAuthToken(oauthParams) {
+  var param_token = oauthParams[0];
+  var param_state = oauthParams[1];
```
It might be clearer to pass oauth params around as an object instead of an array. What do you think?

alisdair commented on PDvis/common.js:
```diff
@@ -17,25 +20,54 @@ function getParametersByName(name) {
+  var param_token = oauthParams[0];
+  var param_state = oauthParams[1];
+  var client_state = window.localStorage.getItem('pdvisClientState');
+  if (param_state != client_state) {
```
Try to avoid single-equals comparisons in JS, otherwise accidental coercions can cause sadness-inducing bugs. `==` -> `===`, `!=` -> `!==`.

alisdair commented on PDvis/common.js:
```diff
@@ -17,25 +20,54 @@ function getParametersByName(name) {
 		type: method,
 		dataType: "json",
-		url: "https://api.pagerduty.com/" + endpoint,
+		url: "http://api.pagerduty.net/" + endpoint,
```
Production URLs

alisdair commented on PDvis/common.js:
```diff
@@ -46,12 +78,16 @@ function PDRequest(token, endpoint, method, options) {
 
 			console.log(alertStr);
+
+      console.log("Attempting to get new OAuth token");
```
Probably remove the console logs as well as the debugger statements

alisdair commented on PDvis/vis.js:
```diff
@@ -826,6 +826,18 @@ function setProgressBar(progress) {
 
 function main() {
+	debugger
+	var oauthResponseParams = getOAuthResponseParams()
```
Maybe this should be inside the following `if` statement? I don't think we need to fetch them if the token is present in localstorage

alisdair commented on PDvis/common.js:
```diff
@@ -17,25 +20,54 @@ function getParametersByName(name) {
+	var param_token = getParameterByName('access_token', hash);
+  if (param_token) oauth_params.push(param_token);
+	var param_state = getParameterByName('state', hash);
+  if (param_state) oauth_params.push(param_state);
```
I think we should clear `window.location.hash` at the end of this function, so that the URL bar doesn't have the token sitting in it once we've stored the data.

jamestyack commented:
Nice PR and work on this! Like the way you added those questions in the description (and quickly got useful feedback from Alisdair); seems really efficient.

MichaelViveros commented:
Some questions:

1. What's the best way to store an app's `client_id`? I'm [hardcoding it](https://github.com/PagerDuty/addons/pull/23/files#diff-073f48c76203d07dc39b9549aa405ff4R48) in the JS code right now. SleepyDuty uses an [environment variable](https://github.com/PagerDuty/sleepy-duty/blob/master/config/environment.js#L45) so maybe I should do that.

2. I thought the code for the OAuth flow in [`main`](https://github.com/PagerDuty/addons/pull/23/files#diff-d4165290cddbf3212be36d99822dd88aR828) was confusing initially so I refactored it to use the `requestOAuthToken` and `receiveOAuthToken` methods which I think helped but it's still a bit confusing so I added `OAUTH.md`. Do you think `OAUTH.md` is necessary? If so, do you think it goes into an adequate level of detail (not too much, not too little)?

brhou commented:
@MichaelViveros 

1. The apps client id is not secret so no worries about hardcoding it. I did it just because in sleep-duty and doesn't need to be like that.
2. The OAuth.md has good content but I would check with John to make sure we need to be releasing this information at this time and in this manner. My sense is that we probably want some dedicated page explaining OAuth on our developer docs and then pointing to this addon as an example instead of a .md.

I think we would also want to implement logout here, which would just be to remove the token from the local storage so that the users have the option to revoke their own token. It would be good practice to show other developers how they would allow that.

jbaldo commented:
@iminoso could we just add a note letting followers here know that OAuth is in beta and people can contact me (jbaldo@pagerduty.com) for early access to OAuth registration.

jbaldo commented on PDvis/common.js:
```diff
@@ -17,25 +24,71 @@ function getParametersByName(name) {
+  window.localStorage.setItem('pdvisClientState', state);
+  var clientId = "b5236b4b54e5ec7b2b51cccc603174a2ac0b575f33753e04a2924dced669c03b";
+  var redirectUri = "https://pagerduty.github.io/addons/PDvis/index.html";
+  var oauthRoute = "http://app.pagerduty.com/oauth/authorize?client_id=" + clientId + "&redirect_uri=" + redirectUri + "&response_type=token&state=" + state;
```
Noticing that Google requires people to supply a https URL.

MichaelViveros replied:
Think sleepy-duty uses [https](https://github.com/PagerDuty/sleepy-duty/blob/6fde69e9a9e711dc6319781419b10b46e971b49e/config/environment.js#L71) too. I'll update PDvis to use https.

MichaelViveros replied:
@jbaldo I actually fixed PDvis to use https in a [commit](https://github.com/PagerDuty/addons/commit/814d9cda68aadf445b6259168119d4a3a0867026#diff-073f48c76203d07dc39b9549aa405ff4R50) shortly after this PR was merged. So it currently uses https but this PR diff shows http.

Don't think I can add commits to this PR anymore since it's merged. So probably best to leave this PR as is and just let people know they can use https. If this issue keeps coming up, I can try making another PR that's a copy of this one that uses https.

jbaldo commented:
Oh, ok. Cool. No problem.

On Mon, Dec 11, 2017 at 10:23 AM, Michael Viveros <notifications@github.com>
wrote:

> *@MichaelViveros* commented on this pull request.
> ------------------------------
>
> In PDvis/common.js
> <https://github.com/PagerDuty/addons/pull/23#discussion_r156156494>:
>
> > -		return;
> -	}
> +  for (var i = 0; i < length; i++) {
> +    result += chars.charAt(Math.floor(Math.random() * chars.length));
> +  }
> +
> +  return result;
> +
> +}
> +
> +function requestOAuthToken() {
> +  var state = generateRandomState(16);
> +  window.localStorage.setItem('pdvisClientState', state);
> +  var clientId = "b5236b4b54e5ec7b2b51cccc603174a2ac0b575f33753e04a2924dced669c03b";
> +  var redirectUri = "https://pagerduty.github.io/addons/PDvis/index.html";
> +  var oauthRoute = "http://app.pagerduty.com/oauth/authorize?client_id=" + clientId + "&redirect_uri=" + redirectUri + "&response_type=token&state=" + state;
>
> @jbaldo <https://github.com/jbaldo> I actually fixed PDvis to use https
> in a commit
> <https://github.com/PagerDuty/addons/commit/814d9cda68aadf445b6259168119d4a3a0867026#diff-073f48c76203d07dc39b9549aa405ff4R50>
> shortly after this PR was merged. So it currently uses https but this PR
> diff shows http.
>
> Don't think I can add commits to this PR anymore since it's merged. So
> probably best to leave this PR as is and just let people know they can use
> https. If this issue keeps coming up, I can try making another PR that's a
> copy of this one that uses https.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PagerDuty/addons/pull/23#discussion_r156156494>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABIsZ-LSR7Snf7FLNqR-BKL5wFeXHbpCks5s_XMigaJpZM4QdxpQ>
> .
>
